// @superset-ui/core/esm/chart-composition/tooltip/TooltipTable.js
import _pt from "prop-types";
import React, { PureComponent } from 'react';
const defaultProps = {
  className: '',
  data: []
};
const VALUE_CELL_STYLE = {
  paddingLeft: 8,
  textAlign: 'right'
};
export default class TooltipTable extends PureComponent {
  render() {
    const {
      className,
      data
    } = this.props;
    return /*#__PURE__*/React.createElement("table", {
      className: className
    }, /*#__PURE__*/React.createElement("tbody", null, data.map(({
      key,
      keyColumn,
      keyStyle,
      valueColumn,
      valueStyle
    }) => /*#__PURE__*/React.createElement("tr", {
      key: key
    }, /*#__PURE__*/React.createElement("td", {
      style: keyStyle
    }, keyColumn != null ? keyColumn : key), /*#__PURE__*/React.createElement("td", {
      style: valueStyle ? { ...VALUE_CELL_STYLE,
        ...valueStyle
      } : VALUE_CELL_STYLE
    }, valueColumn)))));
  }

}
TooltipTable.propTypes = {
  className: _pt.string,
  data: _pt.arrayOf(_pt.shape({
    key: _pt.oneOfType([_pt.string, _pt.number]).isRequired,
    keyColumn: _pt.node,
    valueColumn: _pt.node.isRequired
  })).isRequired
};
TooltipTable.defaultProps = defaultProps;
