// @superset-ui/core/lib/chart/components/FallbackComponent.js
"use strict";

exports.__esModule = true;
exports.default = FallbackComponent;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const CONTAINER_STYLE = {
  backgroundColor: '#000',
  color: '#fff',
  overflow: 'auto',
  padding: 32
};

function FallbackComponent({
  componentStack,
  error,
  height,
  width
}) {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: { ...CONTAINER_STYLE,
      height,
      width
    }
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("b", null, "Oops! An error occured!")), /*#__PURE__*/_react.default.createElement("code", null, error ? error.toString() : 'Unknown Error')), componentStack && /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("b", null, "Stack Trace:"), /*#__PURE__*/_react.default.createElement("code", null, componentStack.split('\n').map(row => /*#__PURE__*/_react.default.createElement("div", {
    key: row
  }, row)))));
}
